### DISCUSSION TITLE
How to use this plugin

### DISCUSSION THREAD

**Discussion by StarCoderLn** [nonzzz/vite-plugin-compression#55] [Q&A] [answered]
# Bug report 🐞
When I used this tool to gzip the element-plus package, the compressed file I got was wrong and could not be used.

## Version & Environment
vue3 + vite project

## Expection
I hope the compressed product can be used normally.

## Actual results (or Errors)
This is how I use it:
![image](https://github.com/nonzzz/vite-plugin-compression/assets/31439591/ca047ba5-197c-4282-8596-64b41c771c31)
The results obtained:
![image](https://github.com/nonzzz/vite-plugin-compression/assets/31439591/1d760a1e-bbba-4687-b789-bb83a6d0deae)

**Comment by nonzzz** [comment]
vite's version?

**Comment by nonzzz** [comment]
Can you share your full vite's config or create a minimal reproduction

**Comment by StarCoderLn** [comment]
> vite's version?

"vite": "4.0.0"

**Comment by nonzzz** [comment]
I created a simple scene and it works fine.
![image](https://github.com/nonzzz/vite-plugin-compression/assets/52351095/b46e4f82-b3ac-4ce3-88bd-7415e8101bea)
Here is my config.

```ts
import { defineConfig } from "vite";

import vue from "@vitejs/plugin-vue";
import compression from "vite-plugin-compression2";

export default defineConfig({
  plugins: [vue(), compression({ deleteOriginalAssets: true })],
});


```

**Comment by StarCoderLn** [comment]
> I created a simple scene and it works fine. ![image](https://private-user-images.githubusercontent.com/52351095/347360162-b46e4f82-b3ac-4ce3-88bd-7415e8101bea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA2MDY4NTAsIm5iZiI6MTcyMDYwNjU1MCwicGF0aCI6Ii81MjM1MTA5NS8zNDczNjAxNjItYjQ2ZTRmODItYjNhYy00Y2UzLTg4YmQtNzQxNWU4MTAxYmVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEwVDEwMTU1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY0YjdjN2QyYjUwYWE0OTk0YzdmMTZlN2JkMzdkZWU5MzE4NTI1ODEzNzVhMDY4OTA0MzI3NmFhMWIwOGQ0NzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WqNdJ55X7y0rzzjMS5jmkpkfqvg5CCTFJWIj3ekJ4Ds) Here is my config.
> 
> ```ts
> import { defineConfig } from "vite";
> 
> import vue from "@vitejs/plugin-vue";
> import compression from "vite-plugin-compression2";
> 
> export default defineConfig({
>   plugins: [vue(), compression({ deleteOriginalAssets: true })],
> });
> ```

You can try to import element-plus as needed. I got an error after compressing element-plus.

**Comment by nonzzz** [comment]
My entry point 

```ts

import { createApp } from "vue";
import Ele from "element-plus";
import App from "./app.vue";
import "element-plus/dist/index.css";

createApp(App).use(Ele).mount("#app");

```

```vue

<template>
    <div>
        <el-button>1</el-button>
    </div>
</template>

```

**Comment by StarCoderLn** [comment]
> I created a simple scene and it works fine. ![image](https://private-user-images.githubusercontent.com/52351095/347360162-b46e4f82-b3ac-4ce3-88bd-7415e8101bea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA2MDcxODIsIm5iZiI6MTcyMDYwNjg4MiwicGF0aCI6Ii81MjM1MTA5NS8zNDczNjAxNjItYjQ2ZTRmODItYjNhYy00Y2UzLTg4YmQtNzQxNWU4MTAxYmVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEwVDEwMjEyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzNDFmZTYwYjAzNmRkZGRmYjcxYjk1ZTczYTc1NTVlZTM1NjAyNTUyZDA2OWMwYjhlOWE5ZGM4NGEyODE0NWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kPNo4oen5fFFbKCtr2qTbnXK1TFR37RV8BFi9AfVOwY) Here is my config.
> 
> ```ts
> import { defineConfig } from "vite";
> 
> import vue from "@vitejs/plugin-vue";
> import compression from "vite-plugin-compression2";
> 
> export default defineConfig({
>   plugins: [vue(), compression({ deleteOriginalAssets: true })],
> });
> ```

This is how I use element-plus
```js
import { defineConfig } from "vite";
import vue from "@vitejs/plugin-vue";
import { ElementPlusResolver } from "unplugin-vue-components/resolvers";
import compression from 'vite-plugin-compression2';

export default defineConfig({
  plugins: [
    vue({
        template: {
          compilerOptions: {
            isCustomElement: (tag) => tag.includes('swiper-'),
          }
        }
    }),
    AutoImport({
        resolvers: [ElementPlusResolver()],
    }),
    Components({
        resolvers: [ElementPlusResolver()],
    }),
    compression({
        threshold: 500 * 1024,
        filename: '[path][base]',
        deleteOriginalAssets: true
    }),
  ]
});
```

**Comment by nonzzz** [comment]
Ok, Let me guess. The filename is `[path][base]` So the original javaScript file has been replace as an compressed module. But you haven't using the cloud server like amazon s3. See #31 . 
I need to explain that compression isn't terser or other code compressor. If you're using it you should use it with Nginx or tomcat or other etc together.

**Comment by nonzzz** [comment]
![image](https://github.com/nonzzz/vite-plugin-compression/assets/52351095/1f2e1f78-fffc-417d-a280-0a479549413d)
Anything works well. I think you can try using the follow script to do a simple validate.

```js

const http = require('http')
const path = require('path')
const fs = require('fs')
const sirv = require('sirv')

const defaultWD = process.cwd()

const publicPath = path.join(defaultWD, 'dist')

const assets = sirv(publicPath, { gzip: true })

function createServer() {
  const server = http.createServer()

  server.on('request', (req, res) => {
    assets(req, res, () => {
      res.statusCode = 404
      res.end('File not found')
    })
  })

  server.listen(0, () => {
    const { port } = server.address()
    console.log(`server run on http://localhost:${port}`)
  })
}

function main() {
  if (!fs.existsSync(publicPath)) throw new Error("Please check your're already run 'npm run build'")
  createServer()
}

main()


```

Step:
 - Remove `filename` option in plugin. and set the `include`: 'include: [/\.(js)$/, /\.(css)$/]'
 - npm run build (ensure your output is dist)
 - run this script and view the result in your browser.

**Comment by nonzzz** [comment]
AFAIK, `filename` Option, usually for the needs of amazon S3.

**Comment by StarCoderLn** [comment]
> ![image](https://private-user-images.githubusercontent.com/52351095/347366606-1f2e1f78-fffc-417d-a280-0a479549413d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA2MTM5MjUsIm5iZiI6MTcyMDYxMzYyNSwicGF0aCI6Ii81MjM1MTA5NS8zNDczNjY2MDYtMWYyZTFmNzgtZmZmYy00MTdkLWEyODAtMGE0Nzk1NDk0MTNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEwVDEyMTM0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhOTlkMTlhOTlhNDk2YmNjYjFmYTdhNDIzNzhkZmQxYWM2YzkxNTEyZGY1NmMwZDhkMDE0M2I3OTA0ZGIzYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.k3ijkEHXctWoOmKrtazkv4eulCWdGBlKmhGYlbRp3sM) Anything works well. I think you can try using the follow script to do a simple validate.
> 
> ```js
> const http = require('http')
> const path = require('path')
> const fs = require('fs')
> const sirv = require('sirv')
> 
> const defaultWD = process.cwd()
> 
> const publicPath = path.join(defaultWD, 'dist')
> 
> const assets = sirv(publicPath, { gzip: true })
> 
> function createServer() {
>   const server = http.createServer()
> 
>   server.on('request', (req, res) => {
>     assets(req, res, () => {
>       res.statusCode = 404
>       res.end('File not found')
>     })
>   })
> 
>   server.listen(0, () => {
>     const { port } = server.address()
>     console.log(`server run on http://localhost:${port}`)
>   })
> }
> 
> function main() {
>   if (!fs.existsSync(publicPath)) throw new Error("Please check your're already run 'npm run build'")
>   createServer()
> }
> 
> main()
> ```
> 
> Step:
> 
> * Remove `filename` option in plugin. and set the `include`: 'include: [/.(js)$/, /.(css)$/]'
> * npm run build (ensure your output is dist)
> * run this script and view the result in your browser.

ok，thanks for your help.

**Comment by nonzzz** [comment]
So it not a bug. Just some missing in document?

**Comment by StarCoderLn** [comment]
> So it not a bug. Just some missing in document?

@nonzzz 

After the compression is successful, I found that the resources in the dist directory have been reduced to 274k
![image](https://github.com/user-attachments/assets/2bd1b6b9-343b-4935-80f2-717bb0a2ccc8)

But when the web page is loaded, the displayed resource size is still the uncompressed size. Why is this?
![image](https://github.com/user-attachments/assets/f8f77848-8506-4e63-a75b-2bcc0b541570)

**Comment by nonzzz** [comment]
I'm afraid you're not set the response header in your server. For Nginx you can refer [document](https://nginx.org/en/docs/http/ngx_http_gzip_module.html). Or view [question](https://stackoverflow.com/questions/75856884/i-want-to-compress-my-files-in-vitejs-react). This plugin 
compression on the client to reduce the pressure of the server side. If you're using like vercel or other's free cloud server. You might not need this plugin.

**Comment by StarCoderLn** [comment]
> I'm afraid you're not set the response header in your server. For Nginx you can refer [document](https://nginx.org/en/docs/http/ngx_http_gzip_module.html). Or view [question](https://stackoverflow.com/questions/75856884/i-want-to-compress-my-files-in-vitejs-react). This plugin compression on the client to reduce the pressure of the server side. If you're using like vercel or other's free cloud server. You might not need this plugin.

I started a local server and tested it using the server script you provided last time.

![image](https://github.com/user-attachments/assets/4b5e1a78-52e7-465e-85bc-91ce6d210058)

**Comment by nonzzz** [comment, marked as answer]
:) AFAIK. you can try set `Content-Encoding` visible in your local chrome devtool  network tab. Or check your compressed source response header has the `Content-Encoding` If it has in, It means is work well. 
Here is the document for the compression [doc](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Length)
